### DISCUSSION TITLE
Setting headers for server response

### DISCUSSION THREAD

**Discussion by agawley** [trpc/trpc#1985] [Q&A] [answered]
Seems like I have all the noob questions today!

Anyone know how you can set a custom header for each api-endpoint when using the express adapter.

**Comment by jlalmes** [comment, marked as answer]
Couple of solutions:

1. Use the `responseMeta` function (https://trpc.io/docs/nextjs#responsemeta-callback).
2. Pass your Response object (`res`) into `ctx` with the `createContext` function then use `res.header(...)` inside your resolve fn.

FWIW - I use option 2, example:
```ts
export const createContext = async ({ req, res }: CreateExpressContextOptions) => {
  return { req, res };
};

...

.query('...', {
  ...,
  resolve: async ({ ctx }) => {
    ctx.res.header('x-custom-header', 'hello123');
    ...
  },
}
```
